src/libraries/Microsoft.Extensions.Configuration.Binder/gen/Specs/BindingHelperInfo.cs

```diff
@@ -0,0 +1,236 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using SourceGenerators;
+
+namespace Microsoft.Extensions.Configuration.Binder.SourceGeneration
+{
+    public sealed record BindingHelperInfo
+    {
+        public required ImmutableEquatableArray<string> Namespaces { get; init; }
+        public required bool EmitConfigurationKeyCaches { get; init; }
+
+        public required MethodsToGen_CoreBindingHelper MethodsToGen { get; init; }
+        public required ImmutableEquatableArray<ComplexTypeSpec>? TypesForGen_BindCoreMain { get; init; }
+        public required ImmutableEquatableArray<TypeSpec>? TypesForGen_GetCore { get; init; }
+        public required ImmutableEquatableArray<TypeSpec>? TypesForGen_GetValueCore { get; init; }
+        public required ImmutableEquatableArray<ComplexTypeSpec>? TypesForGen_BindCore { get; init; }
+        public required ImmutableEquatableArray<ObjectSpec>? TypesForGen_Initialize { get; init; }
+        public required ImmutableEquatableArray<ParsableFromStringSpec>? TypesForGen_ParsePrimitive { get; init; }
+
+        internal sealed class Builder(TypeIndex _typeIndex)
+        {
+            private readonly Dictionary<TypeRef, bool> _seenTransitiveTypes = new();
+
+            private MethodsToGen_CoreBindingHelper _methodsToGen;
+            private bool _emitConfigurationKeyCaches;
+
+            private readonly Dictionary<MethodsToGen_CoreBindingHelper, HashSet<TypeSpec>> _typesForGen = new();
+
+            private readonly SortedSet<string> _namespaces = new()
+            {
+                "System",
+                "System.CodeDom.Compiler",
+                "System.Globalization",
+                "System.Runtime.CompilerServices",
+                "Microsoft.Extensions.Configuration",
+            };
+
+            public BindingHelperInfo ToIncrementalValue()
```

Comment:
9.0: reinstate sorting for equatable arrays where possible e.g. changing property order shouldn't cause regeneration.